PYIC-7263: Bulk EVCS migration script

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Bulk EVCS migration script

### Why did it change

This adds a new lamdba to perform the bulk migration of VCs from the tactical store to EVCS.

It steals a lot of logic from the report lambda, and moves some classes to a shared lib.


### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7263](https://govukverify.atlassian.net/browse/PYIC-7263)


[PYIC-7263]: https://govukverify.atlassian.net/browse/PYIC-7263?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7263: Bulk EVCS migration script

This adds a new lamdba to perform the bulk migration of VCs from the
tactical store to EVCS.

It steals a lot of logic from the report lambda, and moves some classes
to a shared lib.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
As written, I think we've conflated the size of the batch (i.e. the group of users that we want to migrate in this run) and the size of the page (i.e. how many we fetch from DynamoDB at a time).

I think we'll probably want to be able to control these independently... so the request is probably more like:
```
{
  "batch": {
    "id:": "some-batch-id",
    "lastEvaluatedKey": "<last-user-from-previous-batch>",
    "remaining": 10000
  },
  "pageSize": 500,
  "parallelism": 4
}
```
So this would attempt to process 10,000 records from the given user, in pages of 500.

An open question might be whether we want batch sizes larger than what can be processed within a single lambda timeout... It might be easiest just to try and avoid it!

Otherwise, we'll need to calculate the remaining users after part-completing a particular batch.

Wynndow replied:
I think I wasn't imagining to have such large batches. Having a page be the batch seemed like a natural fit. Does having more, smaller batches seem like a problem? It would mean that if something needed replaying or rolling back it might be a bit more targeted.
The page value being supplied here is actually the "limit" value - there doesn't seem to be hard upper limit, but not supplying it would mean the max value is used. I think it might be 1MB of data returned...

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+                        String.format(
+                                "Starting bulk migration with exclusiveStartKey: %s",
+                                exclusiveStartKey)));
+        int parallelism = request.parallelism() == 0 ? DEFAULT_PARALLELISM : request.parallelism();
```
I would make `request.parallelism()` an `Integer` (not `int`) - that way it's much more explicit that 'not specified' => `DEFAULT_PARALLELISM`

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+                                "Starting bulk migration with exclusiveStartKey: %s",
+                                exclusiveStartKey)));
+        int parallelism = request.parallelism() == 0 ? DEFAULT_PARALLELISM : request.parallelism();
+        var forkJoinPool = new ForkJoinPool(parallelism);
```
I think strictly we should call `forkJoinPool.shutdown()` in a `finally` block around this - so it's not leaked in the next invocation of the lambda.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+
+        for (var page :
+                reportUserIdentityScanDynamoDataStore.scan(
+                        exclusiveStartKey, request.batchSize(), HASH_USER_ID, USER_ID, IDENTITY)) {
```
We could add filters to this scan, to pre-emptively filter out things like non-P2 and already-migrated identities. It probably won't add that much time to include them, but we shouldn't need to.

We'd just need to make sure we understand what we're adding up - i.e. the 'skipped' audit events are only for 'identities that were deleted/migrated since we ran the report'.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    private void migrateIdentity(ReportUserIdentityItem reportItem, BatchSummary batchSummary) {
+        // Skip any non P2 identities
```
Note that it is possible that the user has subsequently proved a P2 identity on the same account - technically this just means the user didn't have a P2 identity when we last ran the report. In practice I expect that's not very common.

Not a problem, since any new P2 identities will already be in EVCS.

Wynndow replied:
Yeah I figured this - but thought that since we're storing all new identities in EVCS it was a moot point.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/Report.java:
```diff
@@ -0,0 +1,61 @@
+
+@Data
+public class Report {
+    private final Map<String, BatchSummary> batchSummaries;
```
per above, this is probably page summaries, rather than batch - which feels less useful to expose

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/Report.java:
```diff
@@ -0,0 +1,61 @@
+        this.lastEvaluatedHashUserId = lastEvaluatedHashUserId;
+    }
+
+    public Report finalise() {
```
Can we just do this in `addBatchSummary` instead?

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/Report.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.Map;
+
+@Data
```
Should possibly be `@Getter` if you only want people to update it using the defined methods

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+                // Send failed audit event
+                batchSummary.incrementFailedTacticalWrite(reportItem.getHashUserId());
+            }
+        } else {
```
Not that important, but I think it might be clearer to arrange so all the skips come first, and then the default is to migrate
- If < P2 -> skipped for being incomplete
- If no VCs -> skipped because no VCs
- If already migrated -> skipped because already migrated
- Otherwise good to go

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+        // The last evaluated key with the number of items in the page - should be enough to
+        // process the same batch in future
+        return String.format(
+                "%s:%d", lastEvaluatedKey == null ? "START" : lastEvaluatedKey, pageSize);
```
I think it might be better to pass in the `batchId`, that way we can decide later what the format is?

Wynndow replied:
This goes back to the question at the top - my mental model is to give the script some parameters and let it decide the batches.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/BatchSummary.java:
```diff
@@ -0,0 +1,55 @@
+        total++;
+    }
+
+    public synchronized void incrementSkipped() {
```
If we're recording this (which seems like a good idea :) ) - perhaps we should also break down the different reasons for skipping, like we do with the failures?

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+        evcsClient.storeUserVCs(
+                userId,
+                vcs.stream()
+                        .filter(vc -> vc.getMigrated() == null)
```
This might need thinking about - it feels a bit iffy to migrate a partial identity as CURRENT, since that should never have happened.

The only valid scenario for an incomplete migration would have been a pending F2F identity, where one half got migrated but the other did not. FWIW we didn't see any of these when running the original report, so this may be an entirely moot point.

Wynndow replied:
We could add partiallly migrated identities to the skipped reasons, and keep track of them. That way if we do encounter any we can decide what to do after?

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,271 @@
+                                vc -> {
+                                    vc.setMigrated(timestamp);
+                                    return vc;
+                                })
```
IntelliJ helpfully tells me this can be simplified with a `peek`:
```
vcs.stream()
        .filter(vc -> vc.getMigrated() == null)
        .peek(vc -> vc.setMigrated(timestamp))
        .toList());
```

Wynndow replied:
That was my original implementation, but then sonar shouted at me for using a method that's supposed to be only meant for debugging. Apparently it can be skipped by the JVM for optimisation reasons. Although if that'd actually happen when you've got it doing "stuff" I'm not sure.

Joe-Edwards-GDS replied:
It sounds like the Sonar rule is slightly misguided, the point is that streams are lazy, and intermediate operations will only occur when they are needed (rather than them being dropped arbitrarily). For example:
```
Stream.of("foo", "bar", "baz")
    .peek((v) -> System.out.println(v))
    .findFirst();
```
This will (probably) only print the first entry, but the same would apply if it was done in a `.map` as well! (They are both intermediate operations).

The additional warning applies to `peek`, because you're more likely to do something side-effecty. (With a `map`, this behaviour is probably a bit more expected)

Joe-Edwards-GDS replied:
In this case, the optimal implementation is probably to use neither of these options:
```
vcs.forEach(vc -> vc.setMigrated(timestamp));
verifiableCredentialService.updateIdentity(vcs);
```
😄

Wynndow committed:
PYIC-7263: Bulk migration script improvemnts

Including:

- Using Integer for parallelism to be more explicit
- Shutting down the forkpool to avoid leak into next lamda invocation
- Using Getter instead of Data to maintain minimal access
- Reordering logic to handle all skips first
- Keeping track of the counts of different skip reasons

Wynndow committed:
PYIC-7263: Allow batches to be passed in to bulk migration script

Allows the definition of a batch to run to be passed in. Also refactors
the logging to be more informative and correct.

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,367 @@
+                    throw new StopBeforeLambdaTimeoutException();
+                }
+
+                if (report.getTotalEvaluated() >= request.batch().size()) {
```
Does this mean we might 'overshoot' a batch, if the fetched page sizes don't line up?

Do you think we need to worry about this, or at least warn if it happens? In practice, if we make sure page size is a divisor of batch size it should be ok 🤞

Wynndow replied:
I did consider this and thought that it probably doesn't matter all that much, we won't be orders of magnitude out. I also thought it's proobably very likely that the page and batch sizes will line up, unless we decide to do something a bit odd.

Joe-Edwards-GDS replied:
Yeah, I think we can make it work.

(I'd just like to avoid having to explain 'we said batch 1 would be X users, but it actually ended up being Y', if possible!)

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/BatchReport.java:
```diff
@@ -0,0 +1,57 @@
+        this.allFailedEvcsWriteHashUserIds = new ArrayList<>();
+    }
+
+    public synchronized void addPageSummary(PageSummary summary) {
```
I don't think this needs to be synchronized - it looks like it's only called on the main thread

(unlike the page ones, which do)

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,367 @@
+                                    .with(PAGE_ITEM_COUNT, pageSummary.getCount()));
+                    report.finalise(
+                            "Lambda close to timeout", getPageLastEvaluatedHashUserId(page));
+                    throw new StopBeforeLambdaTimeoutException();
```
Could just `break` here? (or `return report`)?

Wynndow replied:
More than one break and Sonar gets angry with you. Could just return report I suppose, which would be tidier

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java:
```diff
@@ -0,0 +1,367 @@
+                                    .with(PAGE_ITEM_COUNT, pageSummary.getCount()));
+                    report.finalise(
+                            "All items in batch processed", getPageLastEvaluatedHashUserId(page));
+                    throw new BatchProcessedException();
```
Ditto - could break/return here?

Joe-Edwards-GDS commented on lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/domain/BatchReport.java:
```diff
@@ -0,0 +1,57 @@
+        this.pageSummaries.add(summary);
+    }
+
+    public void finalise(String exitReason, String nextBatchExclusiveStartKey) {
```
I would avoid `finalise` - it could be confused with the `finalize` method (that is called when an object is about to be garbage collected).

Could just be a `setExitReason` and `setNextBatchExclusiveStartKey`?

Wynndow replied:
I wanted to avoid two calls, but I suppose it doesn't really matter

Joe-Edwards-GDS replied:
Yeah, you could always call `setNextBatchExclusiveStartKey` on every iteration (regardless of termination) if you wanted to reduce duplication, but either way is fine.

Joe-Edwards-GDS reviewed (commented):
Looking pretty good I think :)

I imagine most further changes will come out of testing now...

Wynndow committed:
PYIC-7263: Refacot bulk migration script

Improvments include:

- Removal of unneccessary synchronization
- Removal of unneccessary exeptions
- Not using a finalise method
- Dropping a stream for a forEach
